### PR TITLE
introduce a new V2 endpoint for rapid recommendations

### PR DESCRIPTION
**Pull request by tremes**
# Description

This introduces a new  v2 API with a new endpoint returning a new data format (called as remote configuration here). 

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update
- Refactor (refactoring code, removing useless files)
- Unit tests (no changes in the code)
- Configuration update

## Testing steps

Please describe how the change was tested locally. If, for some reason, the testing was not done or not done fully, please describe what are the testing steps.

## Checklist
* [ ] `make before_commit` passes
* [ ] updated documentation wherever necessary
* [ ] added or modified tests if necessary
* [ ] updated schemas and validators in [insights-data-schemas](https://github.com/RedHatInsights/insights-data-schemas) in case of input/output change

### CODE REVIEW

File: internal/service/endpoints.go [hunk 1]
```diff
@@ -38,6 +38,14 @@ type GatheringRulesResponse struct {
 	Rules   interface{} `json:"rules"`
 }
 
+// RemoteConfigurationResponse structure represents HTTP response with
+// remote configuration content
+type RemoteConfigurationResponse struct {
```
Comment (juandspy) [inline internal/service/endpoints.go:43]: isn't this a duplicate of `RemoteConfiguration` from the repository package?

File: internal/service/handler.go [hunk 2]
```diff
@@ -16,13 +16,21 @@ limitations under the License.
...
 
 // APIPrefix is the prefix used in the console-dot environment
 const APIPrefix = "/api/gathering"
 
-// V1Prefix is the version prefix used in the console-dot environment
-const V1Prefix = "/v1"
+const (
+	// V1Prefix is the version prefix used in the console-dot environment
+	V1Prefix = "/v1"
+	// V1Prefix is the version prefix used for the newer API v2
```
Comment (juandspy) [inline internal/service/handler.go:31]: it should be `V2Prefix`

File: internal/service/handler.go [hunk 3]
```diff
@@ -42,4 +50,8 @@ func (s *Handler) Register(r *mux.Router) {
 	r.Handle(APIPrefix+"/gathering_rules", gatheringRulesEndpoint(s.svc))
 	r.HandleFunc(APIPrefix+V1Prefix+"/openapi.json", serveOpenAPI)
 	r.Handle(APIPrefix+V1Prefix+"/gathering_rules", gatheringRulesEndpoint(s.svc))
+
+	ocpVersion := "TBD"
+	v2Path := fmt.Sprintf("%s%s/%s/gathering_rules", APIPrefix, V2Prefix, ocpVersion)
```
Comment (juandspy) [inline internal/service/handler.go:55]: I'd adapt this code to accept any `ocpVersion`

Comment (juandspy) [reply internal/service/handler.go:55]: then we can handle any exception inside the remoteConfigurationEndpoint

File: internal/service/repository.go [hunk 4]
```diff
@@ -38,6 +39,23 @@ type Rules struct {
 	Version string `json:"version,omitempty"`
 }
 
+// ContainerLogRequests defines a type for requesting container
```
Comment (juandspy) [inline internal/service/repository.go:42]: should be `ContainerLogRequest`

File: internal/service/endpoints.go (hunk 1 above)
Comment (tremes) [reply internal/service/endpoints.go:43]: yep good catch. THanks

File: internal/service/handler.go (hunk 3 above)
Comment (tremes) [reply internal/service/handler.go:55]: Something like `%s%s/{ocpVersion}/gathering_rules` right? I can't then get `r.PathValue("ocpVersion")` for some reason

File: internal/service/repository.go [hunk 5]
```diff
@@ -38,6 +39,23 @@ type Rules struct {
 	Version string `json:"version,omitempty"`
 }
 
+// ContainerLogRequest defines a type for requesting container
+// log data
+type ContainerLogRequest struct {
+	Namespace    string   `json:"namespace"`
+	PodNameRegex string   `json:"pod_name_regex"`
+	Previous     string   `json:"previous,omitempty"`
```
Comment (tremes) [inline internal/service/repository.go:47]: This must be boolean.

Comment (JiriPapousek) [reply internal/service/repository.go]: Changed to boolean.

File: internal/service/repository.go [hunk 6]
```diff
@@ -64,3 +82,19 @@ func (r *Repository) Rules() (*Rules, error) {
 
 	return &rules, nil
 }
+
+// RemoteConfiguration returns a default remote configuration for v2 endpoint
+func (r *Repository) RemoteConfiguration() (*RemoteConfiguration, error) {
+	filepath := "config_default.json" // TODO: Make this depend on OCP version
```
Comment (juandspy) [inline internal/service/repository.go:88]: when is this TODO going to be done?

Comment (JiriPapousek) [reply internal/service/repository.go:88]: After we talk about how the mapping should be done (semantic versions/regexes) with Jan Holeček (after he returns from PTO).

File: internal/service/testdata/v2/rules.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        }
+    ],
+    "container_logs": [
+        {
+            "namespace": "namespace-1",
+            "pod_name_regex": "test regex",
+            "previous": true,
+            "messages": [
+                "first message",
+			    "second message"
```
Comment (juandspy) [inline internal/service/testdata/v2/rules.json:19]: this is extra indented

File: internal/service/repository.go (hunk 6 above)
Comment (juandspy) [reply internal/service/repository.go:88]: please create a task so that we don't forget!

**Review by juandspy — APPROVED** [review]

File: internal/service/repository.go (hunk 6 above)
Comment (JiriPapousek) [reply internal/service/repository.go:88]: It's still part of the initial [CCXDEV-12785](https://issues.redhat.com/browse/CCXDEV-12785), that has been postponed to next sprint (even though it's almost done, except for this thing). I'll leave a comment there.

**Review by joselsegura — APPROVED** [review]

**Review by Bee-lee — COMMENTED** [review]
could there be a possible race condition that could happen with the cache here? https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/blob/58254f677456af5636db30abcd216115b96e1263/internal/service/storage.go#L108
since the `readDataFromPath()` method is called every time a request is made and the cache is filled in case it's empty, is it possible that two requests made in short succession could attempt to write to the same "global" variable `cache`?

**Comment by JiriPapousek** [comment]
> could there be a possible race condition that could happen with the cache here?
> 
> https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/blob/58254f677456af5636db30abcd216115b96e1263/internal/service/storage.go#L108
> 
> 
> since the `readDataFromPath()` method is called every time a request is made and the cache is filled in case it's empty, is it possible that two requests made in short succession could attempt to write to the same "global" variable `cache`?

Unfortunatelly it seems to be possible, yes. Since this issue is present in the master and is only slightly refactored in this PR, I'll create a new task and deal with this problem in a separate PR.

**Pull request merged**